### PR TITLE
Remember pos

### PR DESCRIPTION
**Pull request by RadekVana**
BFW-2442
Define screen init variant
Add methods for screen to work with init variant
Modify screen handler to handle screen init variant
Remember position when returning to home screen
Remember position when nesting in menus
Save some RAM in menus via bit fields

### CODE REVIEW

File: src/gui/ScreenHandler.hpp [hunk 1]
```diff
@@ -25,18 +32,20 @@ class Screens {
...
-    Screens(const ScreenFactory::Creator screen_creator);
+    Screens(screen_node screen_creator);
     Screens(const Screens &) = delete;
     static Screens *instance;
 
 public:
     void Loop(); //call inside guiloop
 
-    void Open(const ScreenFactory::Creator screen_creator); //remember creator and create later
+    void Open(const ScreenFactory::Creator screen_creator); //remember creator with and create later with default initialization parameter (like selected item in menu)
```
Comment (espr14) [inline src/gui/ScreenHandler.hpp:42]:
```suggestion
    void Open(const ScreenFactory::Creator screen_creator); //remember creator and create later with default initialization parameter (like selected item in menu)
```

File: src/gui/screen_home.cpp [hunk 2]
```diff
@@ -216,3 +216,25 @@ bool screen_home_data_t::moreGcodesUploaded() {
     lastUploadCount = total;
     return result;
 }
+
+void screen_home_data_t::InitState(screen_init_variant var) {
+    if (!var.GetPosition())
+        return;
+
+    size_t pos = *(var.GetPosition());
+    if (pos >= sizeof(w_buttons) / sizeof(w_buttons[0]))
```
Comment (espr14) [inline src/gui/screen_home.cpp:225]: What about `std::array`? Then we can use safer `w_buttons.size()`.

File: src/guiapi/include/window_menu.hpp [hunk 3]
```diff
@@ -2,16 +2,18 @@
...
 class window_menu_t : public IWindowMenu {
-    uint8_t index;    /// index of cursor
-    int8_t moveIndex; /// accumulator for cursor changes
-    bool redrawAll;   /// triggers whole menu redraw
-    bool clicked;     /// triggers click redraw (item->Click invalidates whole menu, but we need to know what happend to redraw only nessessary items)
+    uint8_t index;      /// index of cursor
+    int8_t moveIndex;   /// accumulator for cursor changes
+    uint8_t top_index;  /// offset of currently shown item
+    bool redrawAll : 1; /// triggers whole menu redraw
+    bool clicked : 1;   /// triggers click redraw (item->Click invalidates whole menu, but we need to know what happend to redraw only nessessary items)
```
Comment (espr14) [inline src/guiapi/include/window_menu.hpp:16]:
```suggestion
    bool clicked : 1;   /// triggers click redraw (item->Click invalidates whole menu, but we need to know what happened to redraw only necessary items)
```

**Review by espr14 — CHANGES_REQUESTED** [review]

File: src/gui/ScreenHandler.cpp [hunk 4]
```diff
@@ -210,9 +213,10 @@ void Screens::Loop() {
 }
 
 void Screens::InnerLoop() {
+    screen_init_variant variant;
```
[the remark this answers is not in the document]
Comment (RadekVana) [reply src/gui/ScreenHandler.cpp:216]: ok

**Review by espr14 — APPROVED** [review]

**Pull request merged**